README.md
# TriDimRegression

Package to calculate the bidimensional and tridimensional regression between two 2D/3D configurations.

## Installation from Github
```
library("devtools"); install_github("alexander-pastukhov/tridim-regression", dependencies=TRUE)
```

If you want vignettes, us 
```
devtools::install_github("alexander-pastukhov/tridim-regression",
                         dependencies=TRUE,
                         build_vignettes = TRUE)
```

## Using TriDimRegression

You can call the main function either via a formula that specifies dependent and independent variables with the `data` table or by supplying two tables one containing all independent variables and one containing all dependent variables. The former call is
```
euc2 <- fit_transformation(depV1 + depV2 ~ indepV1 + indepV2, NakayaData, 'euclidean')
```
whereas the latter is 
```
euc3 <- fit_transformation_df(Face3D_W070, Face3D_W097, transformation ='translation')
```

See also `vignette("calibration", package="TriDimRegression")` for an example of using `TriDimRegression` for 2D eye gaze data and `vignette("comparing_faces", package="TriDimRegression")` for an example of working with 3D facial landmarks data.

For the 2D data, you can fit `"translation"` (2 parameters for translation only), `"euclidean"`
(4 parameters: 2 for translation, 1 for scaling, and 1 for rotation), `"affine"` (6 parameters: 2 for translation and 4 that jointly describe scaling, rotation and sheer), or `"projective"` (8 parameters: affine plus 2 additional parameters to account for projection). For 3D data, you can fit `"translation"` (3 for translation only), `"euclidean_x"`, `"euclidean_y"`, `"euclidean_z"` (5 parameters: 3 for translation scale, 1 for rotation, and 1 for scaling), `"affine"` (12 parameters: 3 for translation and 9 to account for scaling, rotation, and sheer), and `"projective"` (15 parameters: affine plus 3 additional parameters to account for projection). transformations. For details on how matrices are constructed, see `vignette("transformation_matrices", package="TriDimRegression")`.

Once the data is fitted, you can extract the transformation coefficients via `coef()` function and the matrix itself via `transformation_matrix()`. Predicted data, either based on the original data or on the new data, can be generated via `predict()`. Bayesian R-squared can be computed with or without adjustment via `R2()` function. In all three cases, you have choice between summary (mean + specified quantiles) or full posterior samples. `loo()` and `waic()` provide corresponding measures that can be used for comparison via `loo::loo_compare()` function.

## References

-   Tobler, W. R. (1965). Computation of the corresponding of geographical patterns. Papers of the Regional Science Association, 15, 131-139.
-   Tobler, W. R. (1966). Medieval distortions: Projections of ancient maps. Annals of the Association of American Geographers, 56(2), 351-360.
-   Tobler, W. R. (1994). Bidimensional regression. Geographical Analysis, 26(3), 187-212.
-   Friedman, A., & Kohler, B. (2003). Bidimensional regression: Assessing the configural similarity and accuracy of cognitive maps and other two-dimensional data sets. Psychological Methods, 8(4), 468-491.
-   Nakaya, T. (1997). Statistical inferences in bidimensional regression models. Geographical Analysis, 29(2), 169-186.
-   Waterman, S., & Gordon, D. (1984). A quantitative-comparative approach to analysis of distortion in mental maps. Professional Geographer, 36(3), 326-337.

## License

All code is licensed under the [GPL 3.0](https://opensource.org/licenses/GPL-3.0) license.


NEWS.md
# TriDimRegression 1.0.0.0
## First CRAN Release
* Initial CRAN Release

# TriDimRegression 1.0.1
## Bug fixes
* Unbalanced chunk delimiters in package vignettes


R/data.R
#' Nakaya (1997)
#'
#' @description Nakaya, T. (1997) Statistical inferences in bidimensional regression models. Geographical Analysis, 29(2), 169-186.
#'
#' @source \doi{10.1111/j.1538-4632.1997.tb00954.x}
#' @format A data frame with 19 observations on the following 4 variables:
#' \describe{
#'   \item{depV1, depV2}{numeric vectors, dependent variables}
#'   \item{indepV1, indepV2}{numeric vectors, independent variables}
#' }
"NakayaData"

#' Eye gaze calibration data
#'
#' A dataset containing a monocular eye gaze recording with calibration sequence.
#' Courtesy of \href{https://www.uni-bamberg.de/entwicklungspsychologie/transfer/babyforschung-bambi/}{Bamberger Baby Institut: BamBI}.
#'
#' @source \url{https://www.uni-bamberg.de/entwicklungspsychologie/transfer/babyforschung-bambi/}.
#' @format A data frame with 365 rows and 6 variables:
#' \describe{
#'   \item{time}{sample timestamp, in milliseconds}
#'   \item{x, y}{recorded gaze, in internal eye tracker units}
#'   \item{target_x, target_y}{location of the calibration target on the screen, in pixels}
#'   \item{target}{index of the target within the sequence}
#' }
"EyegazeData"


#' Friedman & Kohler (2003), data set #1
#'
#' @description Data from Friedman, A., & Kohler, B. (2003). Bidimensional regression: Assessing
#' the configural similarity and accuracy of cognitive maps and other two-dimensional data sets.
#' Psychological Methods, 8(4), 468-491. DOI: 10.1037/1082-989X.8.4.468
#'
#' @source \doi{10.1037/1082-989X.8.4.468}
#' @format A data frame with 4 observations on the following 4 variables:
#' \describe{
#'   \item{depV1, depV2}{numeric vectors, dependent variables}
#'   \item{indepV1, indepV2}{numeric vectors, independent variables}
#' }
"FriedmanKohlerData1"

#' Friedman & Kohler (2003), data set #2
#'
#' @description Data from Friedman, A., & Kohler, B. (2003). Bidimensional regression: Assessing
#' the configural similarity and accuracy of cognitive maps and other two-dimensional data sets.
#' Psychological Methods, 8(4), 468-491. DOI: 10.1037/1082-989X.8.4.468
#'
#' @source \doi{10.1037/1082-989X.8.4.468}
#' @format A data frame with 4 observations on the following 4 variables:
#' \describe{
#'   \item{depV1, depV2}{numeric vectors, dependent variables}
#'   \item{indepV1, indepV2}{numeric vectors, independent variables}
#' }
"FriedmanKohlerData2"


#' Carbon, C. C. (2013), data set #1
#'
#' @description Example 1 from the domain of aesthetics to show how the method can be utilized
#' for assessing the similarity of two portrayed persons, actually the Mona Lisa in the world
#' famous Louvre version and the only recently re-discovered Prado version.
#'
#' @source \doi{10.18637/jss.v052.c01}
#' @format A data frame with 36 observations on the following 4 variables:
#' \describe{
#'   \item{depV1, depV2}{numeric vectors, dependent variables}
#'   \item{indepV1, indepV2}{numeric vectors, independent variables}
#' }
"CarbonExample1Data"

#' Carbon, C. C. (2013), data set #2
#'
#' @description Example 2 originates from the area of geography and inspects the accuracy of
#' different maps of the city of Paris which were created over the last 350 years as compared
#' to a recent map.
#'
#' @source \doi{10.18637/jss.v052.c01}
#' @format A data frame with 13 observations on the following 4 variables:
#' \describe{
#'   \item{depV1, depV2}{numeric vectors, dependent variables}
#'   \item{indepV1, indepV2}{numeric vectors, independent variables}
#' }
"CarbonExample2Data"

#' Carbon, C. C. (2013), data set #3
#'
#' @description Example 3 focuses on demonstrating how good a cognitive map recalculated from
#' averaged cognitive distance data fits with a related real map.
#'
#' @source \doi{10.18637/jss.v052.c01}
#' @format A data frame with 10 observations on the following 4 variables:
#' \describe{
#'   \item{depV1, depV2}{numeric vectors, dependent variables}
#'   \item{indepV1, indepV2}{numeric vectors, independent variables}
#' }
"CarbonExample3Data"

#' Face landmarks, female, #070
#'
#' @source Carbon, C. C. (2012). The Bamberg DADA Face Database (BaDADA). A standardized high quality Face Database with faces of Different Affective states from Different Angles. Unpublished databank. University of Bamberg, Bamberg.
#' @format A data frame with 64 landmarks on the following 3 variables:
#' \describe{
#'   \item{x, y, z}{numeric vectors, coordinates of face landmarks}
#' }
"Face3D_W070"

#' Face landmarks, female, #097
#'
#' @source Carbon, C. C. (2012). The Bamberg DADA Face Database (BaDADA). A standardized high quality Face Database with faces of Different Affective states from Different Angles. Unpublished databank. University of Bamberg, Bamberg.
#' @format A data frame with 64 landmarks on the following 3 variables:
#' \describe{
#'   \item{x, y, z}{numeric vectors, coordinates of face landmarks}
#' }
"Face3D_W097"

#' Face landmarks, female, #182
#'
#' @source Carbon, C. C. (2012). The Bamberg DADA Face Database (BaDADA). A standardized high quality Face Database with faces of Different Affective states from Different Angles. Unpublished databank. University of Bamberg, Bamberg.
#' @format A data frame with 64 landmarks on the following 3 variables:
#' \describe{
#'   \item{x, y, z}{numeric vectors, coordinates of face landmarks}
#' }
"Face3D_W182"


#' Face landmarks, female, #243
#'
#' @source Carbon, C. C. (2012). The Bamberg DADA Face Database (BaDADA). A standardized high quality Face Database with faces of Different Affective states from Different Angles. Unpublished databank. University of Bamberg, Bamberg.
#' @format A data frame with 64 landmarks on the following 3 variables:
#' \describe{
#'   \item{x, y, z}{numeric vectors, coordinates of face landmarks}
#' }
"Face3D_W243"


#' Face landmarks, male, #010
#'
#' @source Carbon, C. C. (2012). The Bamberg DADA Face Database (BaDADA). A standardized high quality Face Database with faces of Different Affective states from Different Angles. Unpublished databank. University of Bamberg, Bamberg.
#' @format A data frame with 64 landmarks on the following 3 variables:
#' \describe{
#'   \item{x, y, z}{numeric vectors, coordinates of face landmarks}
#' }
"Face3D_M010"

#' Face landmarks, male, #092
#'
#' @source Carbon, C. C. (2012). The Bamberg DADA Face Database (BaDADA). A standardized high quality Face Database with faces of Different Affective states from Different Angles. Unpublished databank. University of Bamberg, Bamberg.
#' @format A data frame with 64 landmarks on the following 3 variables:
#' \describe{
#'   \item{x, y, z}{numeric vectors, coordinates of face landmarks}
#' }
"Face3D_M92"

#' Face landmarks, male, #101
#'
#' @source Carbon, C. C. (2012). The Bamberg DADA Face Database (BaDADA). A standardized high quality Face Database with faces of Different Affective states from Different Angles. Unpublished databank. University of Bamberg, Bamberg.
#' @format A data frame with 64 landmarks on the following 3 variables:
#' \describe{
#'   \item{x, y, z}{numeric vectors, coordinates of face landmarks}
#' }
"Face3D_M101"

#' Face landmarks, male, #244
#'
#' @source Carbon, C. C. (2012). The Bamberg DADA Face Database (BaDADA). A standardized high quality Face Database with faces of Different Affective states from Different Angles. Unpublished databank. University of Bamberg, Bamberg.
#' @format A data frame with 64 landmarks on the following 3 variables:
#' \describe{
#'   \item{x, y, z}{numeric vectors, coordinates of face landmarks}
#' }
"Face3D_M244"
